### PR TITLE
ISPN-13780 Do not allow shared stores to also be configured with fetc…

### PR DESCRIPTION
**Pull request by wburns**
…h persistent state

https://issues.redhat.com/browse/ISPN-13780

### CODE REVIEW

**Comment by pruivo** [comment]
@wburns Documentation is missing cc @oraNod  :)

**Comment by wburns** [comment]
> @wburns Documentation is missing cc @oraNod :)

Yeah, I am aware. I was working on it in parallel. I created the PR just to get test failures in parallel.

File: documentation/src/main/asciidoc/topics/con_shared_cache_stores.adoc [hunk 1]
```diff
@@ -6,16 +6,14 @@ By default, cache stores are local (`shared="false"`).
 
 * Local cache stores are unique to each node; for example, a file-based cache store that persists data to the host filesystem.
 +
-Local cache stores can fetch state and purge on startup to avoid loading stale entries from persistent storage.
+Local cache stores can use purge on startup to avoid loading stale entries from persistent storage.
```
Comment (oraNod) [inline documentation/src/main/asciidoc/topics/con_shared_cache_stores.adoc:9]: "can use purge" could change to "can purge data"

I'm pretty sure purge is transitive so "purge data" would be the correct form here.

File: documentation/src/main/asciidoc/topics/upgrading.adoc [hunk 2]
```diff
@@ -166,7 +166,6 @@ To use `SingleFileStore` as a file-based cache store with Infinispan 13 or later
 <persistence passivation="false">
   <single-file-store shared="false"
                      preload="true"
-                     fetch-state="true"
                      read-only="false"/>
 </persistence>
 ----
```
Comment (oraNod) [inline documentation/src/main/asciidoc/topics/upgrading.adoc:171]: Could we also add a deprecation notice for `fetch-state` to the upgrade guide?

Also I think it's worth explaining this line from the Jira description: "instead always use the first non shared store that has BULK_READ allowed". I've been looking through the docs and schema but I'm not quite sure what this means as an alternative to `fetch-state`. How do you allow BULK_READ for a non-shared store?

Comment (wburns) [reply documentation/src/main/asciidoc/topics/upgrading.adoc:171]: > Could we also add a deprecation notice for fetch-state to the upgrade guide?

Yes, for sure. I forgot to add it.

> Also I think it's worth explaining this line from the Jira description: "instead always use the first non shared store that has BULK_READ allowed". I've been looking through the docs and schema but I'm not quite sure what this means as an alternative to fetch-state. How do you allow BULK_READ for a non-shared store?

Let me clarify this as BULK_READ is actually an internal thing. But basically any store that implements publishEntries method and isn't write only. All of our stores qualify for this.

File: documentation/src/main/asciidoc/topics/con_shared_cache_stores.adoc (hunk 1 above)
Comment (wburns) [reply documentation/src/main/asciidoc/topics/con_shared_cache_stores.adoc:9]: The problem is purge is not being used a verb here but rather the term is "purge on startup". How do you point out terms like this normally? Should I italicize it?

Comment (oraNod) [reply documentation/src/main/asciidoc/topics/con_shared_cache_stores.adoc:9]: Wrapping in double quotes like you have there is the way to go for a known phrase. Thanks.

File: core/src/main/java/org/infinispan/configuration/cache/AbstractStoreConfigurationBuilder.java [hunk 3]
```diff
@@ -55,10 +54,12 @@ public AbstractStoreConfigurationBuilder(PersistenceConfigurationBuilder builder
 
    /**
     * {@inheritDoc}
+    *
+    * @deprecated Deprecated since 14.0. There is no replacement. First non shared store is picked instead.
     */
+   @Deprecated
    @Override
    public S fetchPersistentState(boolean b) {
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/configuration/cache/AbstractStoreConfigurationBuilder.java:62]: I would deprecate the method in `StoreConfigurationChildBuilder`.

File: core/src/main/java/org/infinispan/configuration/cache/Configurations.java [hunk 4]
```diff
@@ -49,19 +49,25 @@ public static boolean isClustered(GlobalConfiguration globalConfiguration) {
       return globalConfiguration.transport().transport() != null;
    }
 
+   /**
+    * @param storeConfiguration
+    * @return
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/configuration/cache/Configurations.java:54]: fill or remove these lines.

File: core/src/main/java/org/infinispan/configuration/cache/PersistenceConfiguration.java [hunk 5]
```diff
@@ -69,12 +69,11 @@ public int connectionInterval() {
    /**
     * Loops through all individual cache loader configs and checks if fetchPersistentState is set on
     * any of them
+    *
+    * @deprecated since 14.0. This will always return false
     */
+   @Deprecated
    public Boolean fetchPersistentState() {
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/configuration/cache/PersistenceConfiguration.java:76]: wondering why it returns a `Boolean` object...

File: core/src/test/java/org/infinispan/statetransfer/StateTransferCacheLoaderFunctionalTest.java [hunk 6]
```diff
@@ -35,7 +35,7 @@ protected EmbeddedCacheManager createCacheManager(String cacheName) {
       // increment the DIMCS store id
       DummyInMemoryStoreConfigurationBuilder dimcs = new DummyInMemoryStoreConfigurationBuilder(configurationBuilder.persistence());
       dimcs.storeName("store number " + id++);
-      dimcs.fetchPersistentState(true).shared(sharedCacheLoader.get()).preload(true);
+      dimcs.fetchPersistentState(!sharedCacheLoader.get()).shared(sharedCacheLoader.get()).preload(true);
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/statetransfer/StateTransferCacheLoaderFunctionalTest.java:38]:
```suggestion
      dimcs.shared(sharedCacheLoader.get()).preload(true);
```

**Review by pruivo — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/infinispan/configuration/cache/AbstractStoreConfigurationBuilder.java (hunk 3 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/configuration/cache/AbstractStoreConfigurationBuilder.java:62]: Sure.

File: core/src/main/java/org/infinispan/configuration/cache/PersistenceConfiguration.java (hunk 5 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/configuration/cache/PersistenceConfiguration.java:76]: A wonder I hope I never need to find out :) But in all seriousness, guessing just a mistake.

**Review by pruivo — APPROVED** [review]

**Review by oraNod — APPROVED** [review]
LGTM

**Pull request merged**